### PR TITLE
chore: remove support for Python 3.7; prefer 3.8

### PR DESCRIPTION
**Pull request by wes-turner**
Python 3.7 is EOL.

* Some docs that referred to 3.7 now refer to 3.8
* Training API "how to" doc no longer mentions any 3.7 docker images
* `CONTRIBUTING.md` no longer says to upgrade pip when installing determined
* `CONTRIBUTING.md` has revised text about Python requirements
* mypy tests against 3.8.
* Features new to 3.8 can be used without suppressing type checking or conditional imports. These changes have been made.
* No longer run `test-cli` in CircleCI against Python 3.7.
* There are no wheels for TF1 using Python 3.8. Removing 3.7 support means TF1 cannot be used. In previous commits, we removed support of TF1. We also clean up those references here. In particular:
  - `e2e_tests/tests/command/test_run:test_k8_init_containers` and  `e2e_tests/tests/command/test_run:test_k8_sidecars` now run on TF2 images
  - Removed references to TF1 images in `e2e_tests`.
  - Removed references to TF1 Docker images in `bumpenvs.yaml`

## Test Plan

tests are green

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by rb-determined-ai — APPROVED** [review]
Looks good to me!

There's still some `37` and `3.7` related things in the circleci config though.

**Comment by wes-turner** [comment]
Converting to "draft" pending the removal of support for TF1 (and the TF1 environments that use Python 3.7). Wheels for TF are not available for Python 3.8 (https://pypi.org/project/tensorflow/1.15.5/)

File: docs/interfaces/cli-ug.rst [hunk 1]
```diff
@@ -35,7 +35,7 @@ and upgrade.
  Installation
 **************
 
-The CLI is distributed as a Python wheel package and requires Python >= 3.7. We recommend setting up
+The CLI is distributed as a Python wheel package and requires Python >= 3.8. We recommend setting up
 a `virtualenv <https://virtualenv.pypa.io/en/latest/>`__ and using the ``pip`` utility to install
 ``determined`` into the environment:
```
Comment (tara-det-ai) [inline docs/interfaces/cli-ug.rst:41]: The Command Line Interface (CLI) is distributed in the form of a Python wheel package and requires Python version 3.8 or later. 

(source: [Microsoft Manual of Style](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/l/later))

**Review by tara-det-ai — COMMENTED** [review]
requested edits

File: docs/interfaces/cli-ug.rst (hunk 1 above)
Comment (wes-turner) [reply docs/interfaces/cli-ug.rst]: Done.

**Review by carolinaecalderon — APPROVED** [review]
LGTM

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: .circleci/real_config.yml [hunk 2]
```diff
@@ -12,9 +12,6 @@ orbs:
   win: circleci/windows@5.0.0
 
 executors:
-  python-37:
-    docker:
-      - image: python:3.7-slim-buster
   python-38:
     docker:
       - image: python:3.8-slim-buster
```
Comment (dannysauer) [inline .circleci/real_config.yml:17]: We're still using the Buster images? :D 

They've moved through Bullseye and are now on Bookworm.  This is probably a separate PR, but those buster images haven't been updated for about a year.  docker-library/python#822

**Review by dannysauer — APPROVED** [review]
I didn't do a `grep -R` or anything, but this LGTM.

File: .circleci/real_config.yml (hunk 2 above)
Comment (wes-turner) [reply .circleci/real_config.yml:17]: https://github.com/determined-ai/determined/pull/8943

**Comment by rb-determined-ai** [comment]
yay, finally!

**Pull request merged**